internal/pkg/template/override/rule.go

```diff
@@ -3,36 +3,198 @@
+	// Value       *yaml.Node
+}
+
+func (r Rule) validate() error {
+	return nil
 }
 
-type ruleNode struct {
-	// next *ruleNode
+func (r Rule) parse() (contentUpserter, error) {
+	return nil, nil
+}
+
+// upsertNode represents a node that needs to be upserted.
+type upsertNode struct {
+	key           string
+	valueToInsert *yaml.Node
+	next          contentUpserter
+}
+
+// NextNode returns the next node.
+func (m *upsertNode) NextNode() contentUpserter {
+	return m.next
 }
 
-func parseRules(rules []Rule) ([]*ruleNode, error) {
-	var ruleNodes []*ruleNode
-	for _, r := range rules {
-		if err := r.validate(); err != nil {
-			return nil, err
+// mapUpsertNode represents a map node that needs to be upserted at the given key.
+type mapUpsertNode struct {
+	upsertNode
+}
+
+// Upsert upserts a node into given yaml content.
+func (m *mapUpsertNode) Upsert(content *yaml.Node) (*yaml.Node, error) {
+	// If it contains the value to insert, upsert the value to the yaml.
+	if m.valueToInsert != nil {
+		m.upsertValue(content)
+		return nil, nil
+	}
+	for i := 0; i < len(content.Content); i += 2 {
+		// The content of a map always come in pairs. If the node pair exists, return the map node.
+		// Note that the rest of code massively uses yaml node tree.
+		// Please refer to https://www.efekarakus.com/2020/05/30/deep-dive-go-yaml-cfn.html
+		if content.Content[i].Value == m.key {
+			return content.Content[i+1], nil
 		}
-		node, err := r.parse()
-		if err != nil {
-			return nil, err
+	}
+	// If the node pair doesn't exist, create the label node first and then a map node.
+	// Finally we return the created map node.
+	newLabelNode := &yaml.Node{
+		Kind:  yaml.ScalarNode,
+		Tag:   nodeTagStr,
+		Value: m.key,
+	}
+	content.Content = append(content.Content, newLabelNode)
+	newValNode := &yaml.Node{
+		Kind: yaml.MappingNode,
+		Tag:  nodeTagMap,
+	}
+	content.Content = append(content.Content, newValNode)
+	return newValNode, nil
+}
+
+func (m *mapUpsertNode) upsertValue(content *yaml.Node) {
+	// If the node pair exists, substitute with the value node.
+	for i := 0; i < len(content.Content); i += 2 {
+		if m.key == content.Content[i].Value {
+			content.Content[i+1] = m.valueToInsert
 		}
-		ruleNodes = append(ruleNodes, node)
 	}
-	return ruleNodes, nil
+	// Otherwise, we create the label node then append the value node.
+	newLabelNode := &yaml.Node{
+		Kind:  yaml.ScalarNode,
+		Tag:   nodeTagStr,
+		Value: m.key,
+	}
+	content.Content = append(content.Content, newLabelNode)
+	content.Content = append(content.Content, m.valueToInsert)
 }
 
-func (r Rule) validate() error {
-	return nil
+// seqIdxUpsertNode represents a sequence node that needs to be upserted at index.
+type seqIdxUpsertNode struct {
+	index        int
+	appendToLast bool
+	upsertNode
 }
 
-func (r Rule) parse() (*ruleNode, error) {
-	return nil, nil
+// Upsert upserts a node into given yaml content.
+func (s *seqIdxUpsertNode) Upsert(content *yaml.Node) (*yaml.Node, error) {
```

Comment:
What do you think of changing `content` to `parentContent`?  
If I'm following correctly, let's say we are given `a.b[0]` over here `content` refers to the Content of `a` right?

Reply:
Yeah i'll change the `mapUpsertNode` as well